framework/include/controls/ControlInterface.h

```diff
@@ -0,0 +1,219 @@
+   *
+   */
+  std::deque<std::string> tokenizeName(const std::string & name);
+
+  /// Reference to the Control objects parameters
+  const InputParameters & _ci_parameters;
+
+  /// A Reference to MooseApp (needed to gain access to InputParameterWarehouse)
+  MooseApp & _ci_app;
+
+  /// A reference to the InputParameterWarehouse which is used for access the parameter objects
+  InputParameterWarehouse & _input_parameter_warehouse;
+
+  /// The object thread id, needed for access the warehouse
+  THREAD_ID _tid;
+
+  /// The name of the object, for better error message
+  const std::string & _ci_name;
+
+  // Only these objects are allowed to construct this interface
+  friend class RealParameterReporter;
+  friend class Control;
+};
+
+template<typename T>
+T &
+ControlInterface::getControlParam(const std::string & name)
+{
+  return getControlParamByName<T>(_ci_parameters.get<std::string>(name));
+}
+
+template<typename T>
+T &
+ControlInterface::getControlParamByName(const std::string & name)
+{
+  // Get the InputParameter objects that contain objects
+  std::deque<std::string> desired = tokenizeName(name);
+  std::vector<InputParameterIterator> iters = getValidInputParameterIteratorsByName<T>(name, desired);
+
+  // If more than one parameter is located, report an eror message listing the object and parameter names
+  if (iters.size() > 1)
+  {
+    std::ostringstream oss;
+    oss << "The controlled parameter, '" << name << "', in " << _ci_name << " was found in multiple objects:\n";
+    for (std::vector<InputParameterIterator>::const_iterator it = iters.begin(); it != iters.end(); ++it)
+      oss << "  " << (*it)->first << '\n';
+    oss << "\nUse the method 'getControlParamVector' to access multiple parameters";
+    mooseError(oss.str());
+  }
+
+  // Return a writeable reference to the parameter
+  return iters[0]->second->set<T>(desired[2]);
+}
+
+template<typename T>
+std::vector<T*>
+ControlInterface::getControlParamVector(const std::string & name)
+{
+  return getControlParamVectorByName<T>(_ci_parameters.get<std::string>(name));
+}
+
+template<typename T>
+std::vector<T*>
+ControlInterface::getControlParamVectorByName(const std::string & name)
+{
+  // Get the InputParameter objects that contain objects
+  std::deque<std::string> desired = tokenizeName(name);
+  std::vector<InputParameterIterator> iters = getValidInputParameterIteratorsByName<T>(name, desired);
+
+  // The vector the return
+  std::vector<T*> output;
+
+  // Loop over all InputParameter objects
+  for (std::vector<InputParameterIterator>::iterator it = iters.begin(); it != iters.end(); ++it)
+    output.push_back(&((*it)->second->set<T>(desired[2])));
+
+  return output;
+}
+
+template<typename T>
+std::vector<InputParameterIterator>
+ControlInterface::getValidInputParameterIteratorsByName(const std::string & name, const std::deque<std::string> & desired)
+{
+  // The input deque consists of:
+  //   desired[0] = system; desired[1] = group; desired[2] = param
```

Comment:
The enum is a good idea, I can implement this for this patch. But, this will all change when I re-factor the object naming. This stuff becomes a struct() with "system", "group", etc. attributes.